branches/SDSC/source/3.2.1/server/r1/utilities/man/man1/Schmod.1
.\" For ascii version, process this file with
.\" groff -man -Tascii Schmod.1
.\"
.TH Schmod 1 "Jan 2003 " "Storage Resource Broker" "User SRB Commands"
.SH NAME
Schmod \- changes permission modes and ownership for objects and collections
in SRB space
.SH SYNOPSIS
.B Schmod
.RB [ \-h ]
.RB [ \-a ]
.RB [ \-p ]
.RB [ \-r ]
.RB [ \-d | c | D]
.RB [ \-C]
.IR "w|a|r|n|t|o|c "
.I newUserName domainName
.IR collection | srbObj " ..."
.SH DESCRIPTION
.B "Schmod "
grants/changes access permissions or changes owners for the operand
.IR collection " or " srbObj " ... for " newUserName " in " domainName .
.sp
If the operand is a dataset (srbObj), the permission is granted
for all copies (replica) of the dataset.
.sp
Access permissions allowed are write (w), read (r), all (a),
annotate (t),  none (n), give curator (c) permission or  change owner (o).
.sp
The user changing permissions should have 'all' access
permission for each SrbObj.
.PP
.SH "OPTIONS"
.TP 0.5i
.B "\-r "
grants/changes access permissions recursively for data and sub-collections in 
.I collection
and in all sub-collections under it.
.TP 0.5i
.B "\-d "
grants/changes access permissions for datasets only.
.TP 0.5i
.B "\-D "
grants/changes access permissions for all datasets and sub-collections under 
.I collection
but does not change any permission for datasets or collections under its sub-collections. With the -r option will do it recursively for sub-collections also.
.TP 0.5i
.B "\-C "
grants/changes access permissions for all datasets and sub-collections under 
.I collection
if the user is a curator for at least one of the ancestor collections. With the -r option will do it recursively for sub-collections also.
.TP 0.5i

.B "\-c "
grants/changes access permissions for collections only.
.TP 0.5i
.B "\-a "
adds auditing feature for the access permission granted
for datasets. Auditing works only if the audit_flag is
turned on when compiling the MCAT Server.
.TP 0.5i
.B "\-p "
prompt user for confirmation before changing.
.TP 0.5i
.B "\-h "
help
.SH "SEE ALSO"
.BR SgetD (1),
.BR SgetU (1)



tags/v3_3_1-1/source/3.3.1/server/r1/utilities/src/Schmod.c
/**************************************************************************
Copyright Notice
All Rights Reserved
Please refer to files in COPYRIGHT directory under the SRB software directory 
for copyright, usage, liability and license information.
Please read these files before using,modifying or distributing SRB software.
**************************************************************************/

#include "scommands.h"


 


extern char mdasCollectionName[];
extern char mdasCollectionHome[];
extern char srbAuth[];
extern char srbHost[];
extern char inCondition[];
extern int  curatorFlag;
extern int  hasPrinted;
void usage (char *prog);
int
main(int argc, char **argv)
{
    int c;
    char mode[MAX_TOKEN];
    int i, status, jj;
    srbPathName nameArray[MAX_TOKEN];
    int nArgv;          /* number of inpArgv */
    int flagval = 0;
    int Dflag = 0;
    int auditFlag = 0;
    char userName[MAX_TOKEN], domainName[MAX_TOKEN], userdom[MAX_TOKEN];
    srbConn *conn;
    int IFlag = 0;
    int collInherit = 0;

    curatorFlag = 0;
    hasPrinted = 0;
    strcpy(inCondition , "");
    while ((c=getopt(argc, argv,"hardpcDCbi:")) != EOF) {
        switch (c) {
            case 'h':
                usage (argv[0]);
		exit (1);
                break;
            case 'r':
                flagval |= R_FLAG;
                break;
            case 'd':
                flagval |= d_FLAG;
                break;
            case 'C':
                curatorFlag = 1;
                break;
            case 'D':
                Dflag = 1;
                break;
	    case 'b':
	        flagval |= R_FLAG;
                Dflag = 1;
                break;

            case 'p':
                flagval |= P_FLAG;
                break;
            case 'c':
                flagval |= C_FLAG;
                break;
            case 'a':
                auditFlag = 1;
                break;
	    case 'i':
                IFlag = 1;
                break;
            default:
                usage (argv[0]);
                exit (1);
        }
    }

    if (IFlag == 0 && argc < optind + 4) {		/* must have at least 4 inputs */
	fprintf(stderr,"Schmod: Must have at least 4 input values\n");
	usage (argv[0]);
	exit (1);
    }
    else if (IFlag == 1 && argc < optind + 2) {              /* must have at least 2 inputs */
        usage (argv[0]);
        exit (1);
    }
    
    if ((flagval & C_FLAG) && (flagval & d_FLAG)) {
        fprintf(stderr,"Schmod:  -d and -c options cannot be both on.\n");
        usage (argv[0]);
	 exit (1);
    }

    if (auditFlag) {
      if (!strcmp(argv[optind], "r"))
        strcpy(mode,"read audit");
      else if (!strcmp(argv[optind], "w"))
        strcpy(mode,"write audit");
      else if (!strcmp(argv[optind], "a"))
        strcpy(mode,"all audit");
      else if (!strcmp(argv[optind], "t"))
        strcpy(mode,"annotate audit");
      else if (!strcmp(argv[optind], "c"))
        strcpy(mode,"curate audit");
      else if (!strcmp(argv[optind], "n"))
        strcpy(mode,"null");
      else if (!strcmp(argv[optind], "o"))
        strcpy(mode,"owner");
      else if (!strcmp(argv[optind], "s"))
        strcpy(mode,"set");
      else {
        fprintf(stderr,"Schmod:  Unknown permission: %s\n", argv[optind]);
        usage (argv[0]);
        exit (1);
      }
    }
    else {
      if (!strcmp(argv[optind], "r"))
        strcpy(mode,"read");
      else if (!strcmp(argv[optind], "w"))
        strcpy(mode,"write");
      else if (!strcmp(argv[optind], "a"))
        strcpy(mode,"all");
      else if (!strcmp(argv[optind], "t"))
        strcpy(mode,"annotate");
      else if (!strcmp(argv[optind], "c"))
        strcpy(mode,"curate");
      else if (!strcmp(argv[optind], "n"))
        strcpy(mode,"null");
      else if (!strcmp(argv[optind], "o"))
        strcpy(mode,"owner");
      else if (!strcmp(argv[optind], "s"))
        strcpy(mode,"set");
      else if (!strcmp(argv[optind], "u"))
        strcpy(mode,"unset");
      else {
        fprintf(stderr,"Schmod:  Unknown permission: %s\n", argv[optind]);
        usage (argv[0]);
        exit (1);
      }
    }
    if (IFlag == 0) {
      strcpy(userName,argv[++optind]);
      strcpy(domainName,argv[++optind]);
      sprintf(userdom,"%s@%s",userName,domainName);
    } else {
	if (strcmp (mode, "set") != 0 && strcmp (mode, "unset") != 0) {
	    fprintf(stderr,"Schmod: mode must be 's' or 'u' for -i option\n");  
            usage (argv[0]);
            exit (1);
	}
    }
    optind++;

    for (i = optind; i < argc; i++) {
	nameArray[i - optind].inpArgv = argv[i];
    }
    nArgv = argc - optind;

   i = initSrbClientEnv();
    if (i < 0)
      {printf("Sls Initialization Error:%i\n",i); exit(1);}

    conn = srbConnect (srbHost, NULL, srbAuth, 
     NULL, NULL, NULL, NULL);
    if (clStatus(conn) != CLI_CONNECTION_OK) {
      fprintf(stderr,"Connection to srbMaster failed.\n");
      fprintf(stderr,"%s",clErrorMessage(conn));
      clFinish(conn); exit(3);
    }

    /* Convert the name to SQL like */
    if (convNameArrayToSQL (nameArray, nArgv) < 0)
      {clFinish(conn); exit(4);}

    /* added for speeding up for D and r options ***/
    if (Dflag || IFlag) {
      if (flagval & R_FLAG) {
	for (i = 0; i < nArgv; i++) {
	  if (curatorFlag == 0) {
	    if (IFlag == 1 ) {
	      if (!strcmp(mode,"set")) 
		status = srbModifyCollect (conn, MDAS_CATALOG, nameArray[i].outArgv,
					   "1","","",C_CHANGE_ACL_INHERITANCE_BIT_RECUR);
	      else
		status = srbModifyCollect (conn, MDAS_CATALOG, nameArray[i].outArgv,
                                           "0","","",C_CHANGE_ACL_INHERITANCE_BIT_RECUR);
	    }
	    else if (!strcmp(mode,"owner"))
	      status = srbModifyCollect (conn, MDAS_CATALOG, nameArray[i].outArgv,
		       userName, domainName,"", D_CHANGE_INCOLL_OWNER_RECUR);
	    else if (!strcmp(mode,"null"))
	      status = srbModifyCollect (conn, MDAS_CATALOG, nameArray[i].outArgv,
		       userName, domainName, mode , D_DELETE_INCOLL_ACCS_RECUR);
	    else
	      status = srbModifyCollect (conn, MDAS_CATALOG, nameArray[i].outArgv,
		        userName, domainName, mode , D_INSERT_INCOLL_ACCS_RECUR);
	  }
	  else { /* curatorFlag = 1 */
	    if (IFlag == 1 ) {
              if (!strcmp(mode,"set"))
                status = srbModifyCollect (conn, MDAS_CATALOG, nameArray[i].outArgv,
                                           "1","","",C_CHANGE_ACL_INHERITANCE_BIT_RECUR_BY_CURATOR);
              else
                status = srbModifyCollect (conn, MDAS_CATALOG, nameArray[i].outArgv,
                                           "0","","",C_CHANGE_ACL_INHERITANCE_BIT_RECUR_BY_CURATOR);
            }
            else if (!strcmp(mode,"owner"))
	      status = srbModifyCollect (conn, MDAS_CATALOG, nameArray[i].outArgv,
		       userName, domainName,"", D_CHANGE_INCOLL_OWNER_RECUR_BY_CURATOR);
	    else if (!strcmp(mode,"null"))
	      status = srbModifyCollect (conn, MDAS_CATALOG, nameArray[i].outArgv,
		       userName, domainName, mode ,D_DELETE_INCOLL_ACCS_RECUR_BY_CURATOR);
	    else
	      status = srbModifyCollect (conn, MDAS_CATALOG, nameArray[i].outArgv,
		        userName, domainName, mode,D_INSERT_INCOLL_ACCS_RECUR_BY_CURATOR);
	  }
	  if (status != 0) {
	    fprintf(stderr,"srbModifyCollect Failed for %s: %i.\n",
		    nameArray[i].outArgv,status);
	    fprintf(stderr,"%s",clErrorMessage(conn));
	    clFinish(conn); 
	    exit(1);
	  }
	}
	clFinish(conn); 
	exit(0);
      }
      else {
	for (i = 0; i < nArgv; i++) {
	  if (curatorFlag == 0) {
	    if (IFlag == 1 ) {
              if (!strcmp(mode,"set"))
                status = srbModifyCollect (conn, MDAS_CATALOG, nameArray[i].outArgv,
                                           "1","","",C_CHANGE_ACL_INHERITANCE_BIT);
	      else
                status = srbModifyCollect (conn, MDAS_CATALOG, nameArray[i].outArgv,
                                           "0","","",C_CHANGE_ACL_INHERITANCE_BIT);
	    }
	    else if (!strcmp(mode,"owner"))
	      status = srbModifyCollect (conn, MDAS_CATALOG, nameArray[i].outArgv,
		  userName, domainName,"", D_CHANGE_INCOLL_OWNER);
	    else if (!strcmp(mode,"null"))
	      status = srbModifyCollect (conn, MDAS_CATALOG, nameArray[i].outArgv,
		   userName, domainName, mode , D_DELETE_INCOLL_ACCS);
	    else
	      status = srbModifyCollect (conn, MDAS_CATALOG, nameArray[i].outArgv,
		   userName, domainName, mode , D_INSERT_INCOLL_ACCS);
	  }
	  else { /* curatorFlag = 1 */
	    if (IFlag == 1 ) {
              if (!strcmp(mode,"set"))
                status = srbModifyCollect (conn, MDAS_CATALOG, nameArray[i].outArgv,
                                           "1","","",C_CHANGE_ACL_INHERITANCE_BIT_BY_CURATOR);
              else
                status = srbModifyCollect (conn, MDAS_CATALOG, nameArray[i].outArgv,
                                           "0","","",C_CHANGE_ACL_INHERITANCE_BIT_BY_CURATOR);
            }
            else if (!strcmp(mode,"owner"))
	      status = srbModifyCollect (conn, MDAS_CATALOG, nameArray[i].outArgv,
		  userName, domainName,"", D_CHANGE_INCOLL_OWNER_BY_CURATOR);
	    else if (!strcmp(mode,"null"))
	      status = srbModifyCollect (conn, MDAS_CATALOG, nameArray[i].outArgv,
		   userName, domainName, mode , D_DELETE_INCOLL_ACCS_BY_CURATOR);
	    else
	      status = srbModifyCollect (conn, MDAS_CATALOG, nameArray[i].outArgv,
		   userName, domainName, mode , D_INSERT_INCOLL_ACCS_BY_CURATOR);
	  }
	  if (status != 0) {
	    fprintf(stderr,"srbModifyCollect Failed for %s: %i.\n",
		    nameArray[i].outArgv,status);
	    fprintf(stderr,"%s",clErrorMessage(conn));
	    clFinish(conn); 
	    exit(1);
	  }
	}
	clFinish(conn); 
	exit(0);
      }
    }
    /* added for speeding up for D and r options ***/


    /* do the chmod */

    jj = queryAndChmod (conn, MDAS_CATALOG, nArgv, nameArray, flagval, mode, userdom);
    if (jj < 0 || hasPrinted ==0)
      exit(1);
    exit(0);
}

int 
queryAndChmod (srbConn *conn, int catType, int nArgv,
srbPathName nameArray[], int flagval, char *mode, char *userdom)
{
    int i;
    int status1 = 0;
    int status = 0;
    char tmpNameStore[MAX_TOKEN * 2];
    for (i = 0; i < nArgv; i++) {
        /* May be it is a dataName */
         strcpy(tmpNameStore,nameArray[i].outArgv);
	if (nameArray[i].type != COLLECTION_T && (flagval & C_FLAG) == 0) {
	    status1 = expandChmodData (conn, catType, nameArray[i].outArgv, 
	      flagval, mode, userdom);
	}
	if ((flagval & d_FLAG) == 0 || (flagval & R_FLAG)) 
	    status = chmodInColl (conn, catType, tmpNameStore, 
	     flagval, mode, userdom);

	if (status1 < 0 && status < 0)
	    fprintf (stderr, "%s: No such dataset or collection\n",
              nameArray[i].inpArgv);
	else
	  hasPrinted = 1;
    }
    return (0);
}

/* expandChmodData - Chmod the dataName including the effects of special char.
 * put the results in dataResult.
 *
 */

int
expandChmodData (srbConn *conn, int catType, char *dataName, int flagval, 
char *mode, char *userdom)
{
    int status;
    mdasC_sql_result_struct dataResult;

   
    status = queryData (conn, catType, dataName, flagval, MAX_GET_RESULTS, 
      &dataResult);

    if (status < 0)
        return status;

    chmodDataByRes (conn, catType, flagval, mode, userdom, &dataResult);

    /* loop until no more answer */

    while (dataResult.continuation_index >= 0) {
        clearSqlResult (&dataResult);
        status = srbGetMoreRows(conn, MDAS_CATALOG,
          dataResult.continuation_index, &dataResult, MAX_GET_RESULTS);
        if (dataResult.row_count == 0  || status != 0)
            return (0);
	chmodDataByRes (conn, catType, flagval, mode, userdom, &dataResult);
    }
    clearSqlResult (&dataResult);
    return (0);
}

void
 usage (char *prog)
{
    fprintf (stderr, "Usage : %s [-h] [-b] [-p] [-a] [-r] [-d|c|D] [-C] w|a|r|n|t|o|c newUserName domainName collection|srbObj ....\n", prog);
    fprintf (stderr, "    w - write;r - read;a - all permission;n - null; t - annotate; c - curate (collection); o - change owner\n");
    fprintf (stderr, "Usage : %s -i [-r] s|u collection ....\n", prog);
    fprintf (stderr, "    s - set inheritance bit for ACL;  r - reset inheritance bit\n");

}


